pkg/cloud/aws/services/certificates/certificates.go

```diff
@@ -69,73 +68,78 @@ type Config struct {
 
-	etcdCAKeyPair, err := getOrGenerateCACert(&s.scope.ClusterConfig.EtcdCAKeyPair, etcdCA)
-	if err != nil {
-		return errors.Wrapf(err, "Failed to generate certs for %q", etcdCA)
+	if !s.scope.ClusterConfig.CAKeyPair.HasCertAndKey() {
+		s.scope.Logger.V(2).Info("Generating keypair for", "user", clusterCA)
+		clusterCAKeyPair, err := generateCACert(&s.scope.ClusterConfig.CAKeyPair, clusterCA)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to generate certs for %q", clusterCA)
+		}
+		s.scope.ClusterConfig.CAKeyPair = clusterCAKeyPair
 	}
-	s.scope.ClusterConfig.EtcdCAKeyPair = etcdCAKeyPair
 
-	fpCAKeyPair, err := getOrGenerateCACert(&s.scope.ClusterConfig.FrontProxyCAKeyPair, frontProxyCA)
-	if err != nil {
-		return errors.Wrapf(err, "Failed to generate certs for %q", frontProxyCA)
+	if !s.scope.ClusterConfig.EtcdCAKeyPair.HasCertAndKey() {
+		s.scope.Logger.V(2).Info("Generating keypair", "user", etcdCA)
+		etcdCAKeyPair, err := generateCACert(&s.scope.ClusterConfig.EtcdCAKeyPair, etcdCA)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to generate certs for %q", etcdCA)
+		}
+		s.scope.ClusterConfig.EtcdCAKeyPair = etcdCAKeyPair
 	}
-	s.scope.ClusterConfig.FrontProxyCAKeyPair = fpCAKeyPair
-
-	saKeyPair, err := getOrGenerateServiceAccountKeys(&s.scope.ClusterConfig.SAKeyPair, serviceAccount)
-	if err != nil {
-		return errors.Wrapf(err, "Failed to generate keyPair for %q", serviceAccount)
+	if !s.scope.ClusterConfig.FrontProxyCAKeyPair.HasCertAndKey() {
+		s.scope.Logger.V(2).Info("Generating keypair", "user", frontProxyCA)
+		fpCAKeyPair, err := generateCACert(&s.scope.ClusterConfig.FrontProxyCAKeyPair, frontProxyCA)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to generate certs for %q", frontProxyCA)
+		}
+		s.scope.ClusterConfig.FrontProxyCAKeyPair = fpCAKeyPair
 	}
-	s.scope.ClusterConfig.SAKeyPair = saKeyPair
 
+	if !s.scope.ClusterConfig.SAKeyPair.HasCertAndKey() {
+		s.scope.Logger.V(2).Info("Generating service account keys", "user", serviceAccount)
+		saKeyPair, err := generateServiceAccountKeys(&s.scope.ClusterConfig.SAKeyPair, serviceAccount)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to generate keyPair for %q", serviceAccount)
+		}
+		s.scope.ClusterConfig.SAKeyPair = saKeyPair
+	}
 	return nil
 }
 
-func getOrGenerateCACert(kp *v1alpha1.KeyPair, user string) (v1alpha1.KeyPair, error) {
-	if kp == nil || !kp.HasCertAndKey() {
-		klog.V(2).Infof("Generating keypair for %q", user)
-		x509Cert, privKey, err := NewCertificateAuthority()
-		if err != nil {
-			return v1alpha1.KeyPair{}, errors.Wrapf(err, "failed to generate CA cert for %q", user)
-		}
-		if kp == nil {
-			return v1alpha1.KeyPair{
-				Cert: EncodeCertPEM(x509Cert),
-				Key:  EncodePrivateKeyPEM(privKey),
-			}, nil
-		}
-		kp.Cert = EncodeCertPEM(x509Cert)
-		kp.Key = EncodePrivateKeyPEM(privKey)
+func generateCACert(kp *v1alpha1.KeyPair, user string) (v1alpha1.KeyPair, error) {
```

Comment:
I changed this function to only do one thing (generate certs) instead of one or the other

Reply:
Might be worthwhile to add a comment / description as part of the change

Reply:
I don't understand the request. I like to reserve comments for when something unexpected is happening, but that's not really the case here. Can you elaborate on what you'd like changed?

Reply:
Oh, I noticed the functions don't have any comments on what they do, just adding a small comment might be useful :)